samples/tacos/src/main/kotlin/com/slack/circuit/tacos/OrderTacosCircuit.kt

```diff
@@ -0,0 +1,356 @@
+  val ingredientsCost: BigDecimal = BigDecimal.ZERO
+)
+
+/** List of wizard steps (in order!) */
+private val orderSteps =
+  persistentListOf(FillingsOrderStep, ToppingsOrderStep, ConfirmationOrderStep, SummaryOrderStep)
+
+internal class OrderTacosPresenter(
+  private val fillingsProducer: OrderStep.StateProducer<FillingsOrderStep.State>,
+  private val toppingsProducer: OrderStep.StateProducer<ToppingsOrderStep.State>,
+  private val confirmationProducer: OrderStep.StateProducer<ConfirmationOrderStep.OrderState>,
+  private val summaryProducer: OrderStep.StateProducer<SummaryOrderStep.SummaryState>,
+) : Presenter<OrderTacosScreen.State> {
+  @Composable override fun present(): OrderTacosScreen.State = presentInternal()
+
+  @Composable
+  internal fun presentInternal(
+    initialStep: OrderStep = FillingsOrderStep,
+    initialOrderDetails: OrderDetails = OrderDetails(),
+  ): OrderTacosScreen.State {
+    var currentStep by remember { mutableStateOf(initialStep) }
+    var orderDetails by remember { mutableStateOf(initialOrderDetails) }
+    var isNextEnabled by remember { mutableStateOf(false) }
+
+    // Generate state for the current order step
+    val stepState =
+      produceState(currentStep, orderDetails) { event ->
+        when (event) {
+          is OrderStep.Validation -> isNextEnabled = event.isOrderValid
+          is OrderStep.UpdateOrder -> orderDetails = updateOrder(event, orderDetails)
+          is OrderStep.Restart -> {
+            currentStep = FillingsOrderStep
+            orderDetails = OrderDetails()
+            isNextEnabled = false
+          }
+        }
+      }
+
+    return OrderTacosScreen.State(
+      headerResId = currentStep.headerResId,
+      orderCost = orderDetails.baseCost + orderDetails.ingredientsCost,
+      stepState = stepState,
+      isPreviousVisible = currentStep.index in 1..2,
+      isNextEnabled = isNextEnabled,
+      isNextVisible = currentStep.index < 2,
+      isFooterVisible = currentStep != SummaryOrderStep
+    ) { navEvent ->
+      // process navigation event from child order step
+      processNavigation(currentStep, navEvent) { currentStep = it }
+    }
+  }
+
+  @Composable
+  private fun produceState(
+    orderStep: OrderStep,
+    orderDetails: OrderDetails,
+    eventSink: (OrderStep.Event) -> Unit
+  ): OrderStep.State =
+    when (orderStep) {
+      is FillingsOrderStep -> fillingsProducer(orderDetails, eventSink)
+      is ToppingsOrderStep -> toppingsProducer(orderDetails, eventSink)
+      is ConfirmationOrderStep -> confirmationProducer(orderDetails, eventSink)
+      is SummaryOrderStep -> summaryProducer(orderDetails, eventSink)
+    }
+}
+
+private fun processNavigation(
+  currentStep: OrderStep,
+  navEvent: OrderTacosScreen.Event,
+  onNavEvent: (OrderStep) -> Unit
+) {
+  val newIndex = currentStep.index + navEvent.indexModifier
+  onNavEvent(orderSteps.getOrElse(newIndex) { currentStep })
```

Comment:
this seems strange to me, is there ever a case where we have an unmatched index?

Reply:
No, not unless there's a programming error. When determining the appropriate next `OrderStep`, I'm calculating a `List` index. I added `getOrElse` simply to avoid the possibility of a out of bounds exception.

Reply:
I kind of think we should error instead. Something something offensive programming rather than defensive. But my LOGAF is low